gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java

```diff
@@ -0,0 +1,131 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.monitoring;
+
+import java.io.IOException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Charsets;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+import kafka.message.MessageAndMetadata;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.metastore.FileContextBasedFsStateStore;
+import org.apache.gobblin.metastore.FileContextBasedFsStateStoreFactory;
+import org.apache.gobblin.metastore.StateStore;
+import org.apache.gobblin.metastore.util.StateStoreCleanerRunnable;
+import org.apache.gobblin.metrics.event.TimingEvent;
+import org.apache.gobblin.runtime.kafka.HighLevelConsumer;
+
+
+/**
+ * A Kafka monitor that tracks {@link org.apache.gobblin.metrics.GobblinTrackingEvent}s reporting statuses of
+ * running jobs. The job statuses are stored as {@link org.apache.gobblin.configuration.State} objects in
+ * a {@link FileContextBasedFsStateStore}.
+ */
+@Slf4j
+public abstract class KafkaJobStatusMonitor extends HighLevelConsumer<byte[], byte[]> {
+  static final String JOB_STATUS_MONITOR_PREFIX = "jobStatusMonitor";
+  public static final String JOB_STATUS_MONITOR_ENABLED_KEY =  JOB_STATUS_MONITOR_PREFIX + ".enabled";
+  public static final String STATE_STORE_TABLE_SUFFIX = "jst";
+
+  static final String JOB_STATUS_MONITOR_TOPIC_KEY = "topic";
+  static final String JOB_STATUS_MONITOR_NUM_THREADS_KEY = "numThreads";
+  static final String JOB_STATUS_MONITOR_CLASS_KEY = "class";
+  static final String DEFAULT_JOB_STATUS_MONITOR_CLASS = KafkaAvroJobStatusMonitor.class.getName();
+
+  private static final String KAFKA_AUTO_OFFSET_RESET_KEY = "auto.offset.reset";
+  private static final String KAFKA_AUTO_OFFSET_RESET_SMALLEST = "smallest";
+  private static final String KAFKA_AUTO_OFFSET_RESET_LARGEST = "largest";
+
+  @Getter
+  private final FileContextBasedFsStateStore<org.apache.gobblin.configuration.State> stateStore;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private static final Config DEFAULTS = ConfigFactory.parseMap(ImmutableMap.of(
+      KAFKA_AUTO_OFFSET_RESET_KEY, KAFKA_AUTO_OFFSET_RESET_SMALLEST));
+
+  public KafkaJobStatusMonitor(String topic, Config config, int numThreads) {
+    super(topic, config.withFallback(DEFAULTS), numThreads);
+    this.stateStore =
+        (FileContextBasedFsStateStore<org.apache.gobblin.configuration.State>) new FileContextBasedFsStateStoreFactory().createStateStore(config,
```

Comment:
Should this be configurable to support DB/ZK state stores?

Reply:
Changed it to be config driven.